crypto: utilize libinvoke for Base64 encoding/decoding

petermax2 opened the pull request:
# Purpose

1. eliminate code duplication between `crypto` and `base64`.
2. export `base64` encoding/decoding functions to all plugins.
3. utilize `libinvoke` within the `crypto` plugin for Base64 encoding/decoding operations.
4. close #1049 .

For full discussion refer to #1049 .

# Checklist

Check relevant points but please do not remove entries.
For docu fixes, spell checking, and similar nothing
needs to be checked.

- [x] commit messages are fine ("module: short statement" syntax and refer to issues)
- [ ] I added unit tests
- [x] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [x] meta data is updated (e.g. README.md of plugins)
- [ ] release notes are updated (doc/news/_preparation_next_release.md)

@markus2330 please review my pull request

# TODO

- [x] proper error propagation (in case `libinvoke` fails to load the plugin, etc.)
- [x] update release notes

markus2330 commented:
Thank you, eliminating code duplication is always a good idea :wink:

petermax2 commented:
jenkins build Dockerfile please

petermax2 commented:
jenkins build Jenkinsfile please

petermax2 commented:
jenkins build all please

petermax2 commented:
@markus2330 Travis stalled (1 Travis job stalled, the others were succesful)  and `elektra-gcc-configure-debian-optimizations` has a dbus problem.

Ready for review and merge!

ingwinlu commented:
til the optimizations script disables optimizations even though the description suggests otherwise.

the node has dbus setup correctly (as demonstrated by the passing of the same tests on the same machine). Maybe some kind of concurrency issue as the test is not setting up dbus for itself (but relies on autoconf).

markus2330 commented:
@ingwinlu Can you please integrate the "no optimization" build configuration in the Jenkinsfile? The job fails quite often.

@petermax2 I am not sure how base64 with crypto works. `infos/needs = base64` suggests that a plugin within the plugin chain does the base64 encoding (which would be a very elegant solution) but the dynamic invocation suggests that it works otherwise. Btw. I think libinvocation already sets an error/warning, an additional error/warning only for "crypto being unable to load base64" seems to me much too specific.

petermax2 commented:
> `infos/needs = base64` suggests that a plugin within the plugin chain does the base64 encoding

Oh, sorry, I misinterpreted the `needs` attribute then. I will fix that.

> I think libinvocation already sets an error/warning

I will check that.

petermax2 commented:
>> `infos/needs = base64` suggests that a plugin within the plugin chain does the base64 encoding
> Oh, sorry, I misinterpreted the needs attribute then. I will fix that.

Fixed.

>> I think libinvocation already sets an error/warning
> I will check that.

You are right! Also fixed.

markus2330 commented:
Thank you for the work!

Can you please update the status of the PR (top post here), it still says TODO regarding the error handling.

And please write a short line in the release notes.

markus2330 commented:
jenkins build all please

markus2330 commented:
Thank you! I am looking forward to the next contribution!
